### PR TITLE
Compiler: clean package

### PR DESCRIPTION
**Pull request by csansoon**
# WIP
Finish cleaning the compiler package.

Changes:
 - Changed the `compile` method to use a recursive approach instead of an iterative one. This makes the code way more readable and organized. The tests have not been changed to check that everything still works as expected.
 - Improved typing for AST nodes
 - Updated `ToolCall` object to contain both function's `name` and `arguments`
 - Added alias
 - Removed unused error definitions

### CODE REVIEW

File: packages/compiler/src/compiler/compile.ts [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
-import parse from '../parser/index'
-import type {
-  BaseNode,
-  EachBlock,
-  ElementTag,
-  IfBlock,
-  TemplateNode,
-} from '../parser/interfaces'
+  TOOL_CALL_TAG,
+} from '$/constants'
```
Comment (andresgutgon) [inline packages/compiler/src/compiler/compile.ts:7]: namespace alias with `$compiler/*` we need this to run the package from nextjs and api apps

File: packages/compiler/vitest.config.ts [hunk 2]
```diff
@@ -1,6 +1,8 @@
+import tsconfigPaths from 'vite-tsconfig-paths'
 import { defineConfig } from 'vitest/config'
 
 export default defineConfig({
+  plugins: [tsconfigPaths()],
```
Comment (andresgutgon) [inline packages/compiler/vitest.config.ts:5]: we might not need rollup. Don't remove yet tho

File: packages/compiler/src/compiler/compile.ts (hunk 1 above)
Comment (csansoon) [reply packages/compiler/src/compiler/compile.ts:7]: What do you mean

File: packages/compiler/vitest.config.ts (hunk 2 above)
Comment (csansoon) [reply packages/compiler/vitest.config.ts:5]: The tests do not configure the alias without this, so it does not find the imported files and it fails.

File: packages/compiler/src/compiler/compile.ts (hunk 1 above)
Comment (andresgutgon) [reply packages/compiler/src/compiler/compile.ts:7]: in tsconfig.ts

File: packages/compiler/vitest.config.ts (hunk 2 above)
Comment (andresgutgon) [reply packages/compiler/vitest.config.ts:5]: ahhh sorry this is vitest. ignore

**Review by andresgutgon — APPROVED** [review]

**Pull request merged**